File: server/tickets.js
'use strict';

const { PATHS } = require('./paths');
const { readJson, writeJson } = require('./store');
const schema = require('./schema');

const MAX_TICKETS = 100000;
const FIELD_LIMIT = 300;

function trimField(value) {
  if (value === null || value === undefined) return '';
  return String(value).trim().slice(0, FIELD_LIMIT);
}

// Minimal RFC 4180 reader: quoted fields, escaped quotes, CRLF or LF rows.
function parseCsv(text) {
  const rows = [];
  let row = [];
  let field = '';
  let inQuotes = false;

  for (let index = 0; index < text.length; index += 1) {
    const char = text[index];

    if (inQuotes) {
      if (char === '"') {
        if (text[index + 1] === '"') {
          field += '"';
          index += 1;
        } else {
          inQuotes = false;
        }
      } else {
        field += char;
      }
      continue;
    }

    if (char === '"') {
      inQuotes = true;
    } else if (char === ',') {
      row.push(field);
      field = '';
    } else if (char === '\n') {
      row.push(field);
      rows.push(row);
      row = [];
      field = '';
    } else if (char !== '\r') {
      field += char;
    }
  }

  row.push(field);
  rows.push(row);

  return rows.filter((entry) => entry.some((cell) => cell.trim() !== ''));
}

/** CSV keeps whatever columns the file declares; the header row names them. */
function readCsv(text) {
  const rows = parseCsv(text);
  if (rows.length === 0) return { columns: [], records: [] };

  const columns = rows[0].map((cell, index) => cell.trim() || `Column ${index + 1}`);
  const keys = columns.map((column, index) => schema.toKey(column, index));

  const records = rows.slice(1).map((cells) =>
    keys.reduce((record, key, index) => ({ ...record, [key]: trimField(cells[index]) }), {})
  );

  return { columns, records };
}

/** JSON keeps the union of every key present, in first-seen order. */
function readJsonUpload(text) {
  const parsed = JSON.parse(text);
  const list = Array.isArray(parsed) ? parsed : parsed && Array.isArray(parsed.tickets) ? parsed.tickets : null;
  if (!list) throw new Error('JSON must be an array of records or an object with a "tickets" array');

  const columns = [];
  const records = list.map((entry) => {
    if (!entry || typeof entry !== 'object') return {};
    return Object.keys(entry).reduce((record, rawKey) => {
      if (rawKey === 'processed') return record;
      const key = schema.toKey(rawKey, columns.length);
      if (!columns.includes(key)) columns.push(key);
      return { ...record, [key]: trimField(entry[rawKey]) };
    }, {});
  });

  return { columns, records };
}

/**
 * Coerces raw records against a schema, reporting what had to be dropped
 * rather than quietly shrinking the pool.
 */
function normalizeRecords(records, ticketSchema, duplicatePolicy = 'skip') {
  const keys = ticketSchema.fields.map((field) => field.key);
  const identifier = ticketSchema.identifier;
  const seen = new Set();
  const accepted = [];
  const issues = { missingIdentifier: 0, duplicates: 0, truncated: false };

  records.forEach((entry) => {
    if (!entry || typeof entry !== 'object') {
      issues.missingIdentifier += 1;
      return;
    }

    const id = trimField(entry[identifier]);
    if (!id) {
      issues.missingIdentifier += 1;
      return;
    }

    const seenKey = id.toLowerCase();
    if (seen.has(seenKey)) {
      issues.duplicates += 1;
      if (duplicatePolicy === 'skip') return;
    }

    if (accepted.length >= MAX_TICKETS) {
      issues.truncated = true;
      return;
    }

    seen.add(seenKey);
    accepted.push(keys.reduce((record, key) => ({ ...record, [key]: trimField(entry[key]) }), {}));
  });

  return { tickets: accepted, issues };
}

/**
 * Parses an uploaded file and derives the schema its columns imply. The
 * caller decides whether to adopt that schema or keep the configured one.
 */
function parseUpload(content, format, previousSchema) {
  const text = String(content || '').trim();
  if (!text) throw new Error('The uploaded file is empty');

  const looksLikeJson = text.startsWith('{') || text.startsWith('[');
  const resolvedFormat = format === 'csv' || format === 'json' ? format : looksLikeJson ? 'json' : 'csv';
  const { columns, records } = resolvedFormat === 'json' ? readJsonUpload(text) : readCsv(text);

  if (columns.length === 0) throw new Error('No columns were found in that file.');

  return { format: resolvedFormat, columns, records, schema: schema.schemaFromColumns(columns, previousSchema) };
}

/**
 * Reads the column names actually present in the participant file, in
 * first-seen order. Used to derive a schema when none is configured yet, so a
 * fresh install adopts whatever shape the organiser's data already has.
 */
function inferSchemaFromFile() {
  const data = readJson(PATHS.tickets, { tickets: [] }) || {};
  const records = Array.isArray(data) ? data : data.tickets || [];
  const columns = [];

  records.slice(0, 50).forEach((record) => {
    if (!record || typeof record !== 'object') return;
    Object.keys(record).forEach((key) => {
      if (key !== 'processed' && !columns.includes(key)) columns.push(key);
    });
  });

  return columns.length > 0 ? schema.schemaFromColumns(columns) : null;
}

function loadTicketSource(ticketSchema, duplicatePolicy = 'skip') {
  const data = readJson(PATHS.tickets, { tickets: [] }) || {};
  const records = Array.isArray(data) ? data : data.tickets || [];
  const { tickets, issues } = normalizeRecords(records, ticketSchema, duplicatePolicy);
  return { tickets, issues, rawCount: records.length };
}

function saveTickets(tickets) {
  writeJson(PATHS.tickets, { tickets });
}

function toCsv(rows, columns) {
  const escape = (value) => {
    const text = value === null || value === undefined ? '' : String(value);
    return /[",\n]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
  };

  return [
    columns.map((column) => escape(column.label)).join(','),
    ...rows.map((row) => columns.map((column) => escape(row[column.key])).join(',')),
  ].join('\n');
}

module.exports = {
  MAX_TICKETS,
  parseCsv,
  parseUpload,
  normalizeRecords,
  inferSchemaFromFile,
  loadTicketSource,
  saveTickets,
  toCsv,
};

File: pyserver/tickets.py
"""Participant file parsing, normalization and export."""

import csv
import io
import json
from typing import Any, Dict, List, Optional, Sequence, Tuple

from . import schema
from .paths import TICKETS_PATH
from .store import read_json, write_json

MAX_TICKETS = 100000
FIELD_LIMIT = 300


def _trim(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()[:FIELD_LIMIT]


def read_csv(text: str) -> Dict[str, Any]:
    """CSV keeps whatever columns the file declares; the header row names them."""
    rows = [row for row in csv.reader(io.StringIO(text)) if any(cell.strip() for cell in row)]
    if not rows:
        return {"columns": [], "records": []}

    columns = [cell.strip() or f"Column {index + 1}" for index, cell in enumerate(rows[0])]
    keys = [schema.to_key(column, index) for index, column in enumerate(columns)]

    records = [
        {key: _trim(cells[index]) if index < len(cells) else "" for index, key in enumerate(keys)}
        for cells in rows[1:]
    ]
    return {"columns": columns, "records": records}


def read_json_upload(text: str) -> Dict[str, Any]:
    """JSON keeps the union of every key present, in first-seen order."""
    parsed = json.loads(text)
    if isinstance(parsed, list):
        items = parsed
    elif isinstance(parsed, dict) and isinstance(parsed.get("tickets"), list):
        items = parsed["tickets"]
    else:
        raise ValueError('JSON must be an array of records or an object with a "tickets" array')

    columns: List[str] = []
    records: List[Dict[str, str]] = []

    for entry in items:
        if not isinstance(entry, dict):
            records.append({})
            continue
        record: Dict[str, str] = {}
        for raw_key, value in entry.items():
            if raw_key == "processed":
                continue
            key = schema.to_key(raw_key, len(columns))
            if key not in columns:
                columns.append(key)
            record[key] = _trim(value)
        records.append(record)

    return {"columns": columns, "records": records}


def normalize_records(
    records: Sequence[Any], ticket_schema: Dict[str, Any], duplicate_policy: str = "skip"
) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
    """Coerce raw records against a schema, reporting what had to be dropped."""
    keys = [field["key"] for field in ticket_schema["fields"]]
    identifier = ticket_schema["identifier"]
    seen = set()
    accepted: List[Dict[str, Any]] = []
    issues = {"missingIdentifier": 0, "duplicates": 0, "truncated": False}

    for entry in records:
        if not isinstance(entry, dict):
            issues["missingIdentifier"] += 1
            continue

        identity = _trim(entry.get(identifier))
        if not identity:
            issues["missingIdentifier"] += 1
            continue

        seen_key = identity.lower()
        if seen_key in seen:
            issues["duplicates"] += 1
            if duplicate_policy == "skip":
                continue

        if len(accepted) >= MAX_TICKETS:
            issues["truncated"] = True
            continue

        seen.add(seen_key)
        accepted.append({key: _trim(entry.get(key)) for key in keys})

    return accepted, issues


def parse_upload(content: Any, fmt: Any = None, previous_schema: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Parses an uploaded file and derives the schema its columns imply."""
    text = str(content or "").strip()
    if not text:
        raise ValueError("The uploaded file is empty")

    looks_like_json = text.startswith("{") or text.startswith("[")
    resolved = fmt if fmt in ("csv", "json") else ("json" if looks_like_json else "csv")
    parsed = read_json_upload(text) if resolved == "json" else read_csv(text)

    if not parsed["columns"]:
        raise ValueError("No columns were found in that file.")

    return {
        "format": resolved,
        "columns": parsed["columns"],
        "records": parsed["records"],
        "schema": schema.schema_from_columns(parsed["columns"], previous_schema or {}),
    }


def infer_schema_from_file() -> Optional[Dict[str, Any]]:
    """Read the column names actually present in the participant file.

    Used to derive a schema when none is configured yet, so a fresh install
    adopts whatever shape the organiser's data already has.
    """
    data = read_json(TICKETS_PATH, {"tickets": []}) or {}
    records = data if isinstance(data, list) else data.get("tickets") or []

    columns: List[str] = []
    for record in records[:50]:
        if not isinstance(record, dict):
            continue
        for key in record:
            if key != "processed" and key not in columns:
                columns.append(key)

    return schema.schema_from_columns(columns) if columns else None


def load_ticket_source(ticket_schema: Dict[str, Any], duplicate_policy: str = "skip") -> Dict[str, Any]:
    data = read_json(TICKETS_PATH, {"tickets": []}) or {}
    records = data if isinstance(data, list) else data.get("tickets") or []
    tickets, issues = normalize_records(records, ticket_schema, duplicate_policy)
    return {"tickets": tickets, "issues": issues, "rawCount": len(records)}


def save_tickets(tickets: Sequence[Dict[str, Any]]) -> None:
    write_json(TICKETS_PATH, {"tickets": list(tickets)})


def to_csv(rows: Sequence[Dict[str, Any]], columns: Sequence[Tuple[str, str]]) -> str:
    buffer = io.StringIO()
    writer = csv.writer(buffer, lineterminator="\n")
    writer.writerow([label for _key, label in columns])
    for row in rows:
        writer.writerow([row.get(key, "") for key, _label in columns])
    return buffer.getvalue()
